gui/src/main/proxy.js

```diff
@@ -0,0 +1,107 @@
+                                  'Dynamically sets whether to always send credentials for HTTP NTLM or Negotiate authentication.')
+  .describe('login-by-realm', 'comma-separated list of realm:user:password')
+  .help('help')
+  .parse()
+
+log.warn({config}, 'argv')
+
+const printCertificate = (certif) => `Certificate(${certif.issuerName} ${certif.subjectName})`
+
+module.exports = (app, session, doneSetup) => {
+  const loginByRealm = {}
+  if (config['login-by-realm']) {
+    config['login-by-realm'].split(',').forEach((lbr) => {
+      const [realm, username, password] = lbr.split(':')
+      loginByRealm[realm] = [username, password]
+    })
+  }
+
+  if (config['proxy-ntlm-domains']) {
+    session.defaultSession.allowNTLMCredentialsForDomains(config['proxy-ntlm-domains'])
+  }
+
+  session.defaultSession.setCertificateVerifyProc((request, callback) => {
+    const {hostname, certificate, verificationResult, errorCode} = request
+    if (verificationResult < 0) {
+      log.warn({hostname, certificate: printCertificate(certificate), verificationResult, errorCode}, 'Certificate Verification Error')
+    } else {
+      log.debug({hostname, certificate: printCertificate(certificate), verificationResult, errorCode}, 'Certificate Validated')
+    }
+    callback(-3) // use chrome validation
+  })
+
+  app.on('select-client-certificate', (event, webContents, url, list, callback) => {
+    log.debug({url}, 'select-client-certificate')
+    callback()
+  })
+
+  app.on('certificate-error', (event, webContents, url, error, certificate, callback) => {
+    log.warn({url, error, certificate: printCertificate(certificate)}, 'App Certificate Error')
+    callback(false)
+  })
+
+  app.on('login', (event, webContents, request, authInfo, callback) => {
+    log.warn({request: request.method + ' ' + request.url, authInfo}, 'Login event')
+    const auth = loginByRealm[authInfo.realm]
+    if (auth) {
+      event.preventDefault()
+      callback(...auth)
+    } else {
+      callback()
+    }
+  })
+
+  const electronFetch = require('electron-fetch')
+  global.fetch = (url, opts = {}) => {
+    opts.session = session.defaultSession
+    return electronFetch(url, opts)
+  }
+  http.Agent.globalAgent = http.globalAgent = https.globalAgent = new ElectronProxyAgent(session.defaultSession)
+  const _httpRequest = http.request
+  http.request = function (options, cb) {
+    log.warn(options, 'USING RAW HTTP REQUEST')
+    options.agent = options.agent || http.globalAgent
+    options.headers = options.headers || {}
+    options.headers.host = options.hostname
+    return _httpRequest.call(http, options, cb)
+  }
+  const _httpsRequest = https.request
+  https.request = function (options, cb) {
+    log.warn(options, 'USING RAW HTTPS REQUEST')
+    if (typeof options === 'string') {
+      options = url.parse(options)
+    } else {
+      options = Object.assign({}, options)
+    }
+    options.agent = options.agent || https.globalAgent
```

Comment:
Why doesn't `https.request` need the same `options.headers.host` patching as in `http.request`?

Reply:
`https.request` calls `http.request` https://github.com/nodejs/node/blob/edb03cb65d51e336713d6af1134a7394c0776635/lib/https.js#L239

Reply:
Ok, thanks for the explanation!